soroban-simulation/src/ledger_storage.rs

```diff
@@ -0,0 +1,212 @@
+use super::state_ttl::{get_restored_ledger_sequence, TTLLedgerEntry};
+use soroban_env_host::storage::SnapshotSource;
+use soroban_env_host::xdr::ContractDataDurability::{Persistent, Temporary};
+use soroban_env_host::xdr::{
+    ConfigSettingEntry, ConfigSettingId, Error as XdrError, LedgerEntry, LedgerEntryData,
+    LedgerKey, LedgerKeyConfigSetting, Limits, ScError, ScErrorCode, WriteXdr,
+};
+use soroban_env_host::HostError;
+use std::cell::RefCell;
+use std::collections::HashSet;
+use std::convert::TryInto;
+use std::ffi::NulError;
+use std::rc::Rc;
+use std::str::Utf8Error;
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("not found")]
+    NotFound,
+    #[error("entry is not live")]
+    NotLive,
+    #[error("xdr processing error: {0}")]
+    Xdr(#[from] XdrError),
+    #[error("nul error: {0}")]
+    NulError(#[from] NulError),
+    #[error("utf8 error: {0}")]
+    Utf8Error(#[from] Utf8Error),
+    #[error("unexpected config ledger entry for setting_id {setting_id}")]
+    UnexpectedConfigLedgerEntry { setting_id: String },
+    #[error("unexpected ledger entry type ({ledger_entry_type}) for ttl ledger key")]
+    UnexpectedLedgerEntryTypeForTtlKey { ledger_entry_type: String },
+}
+
+impl From<Error> for HostError {
```

Comment:
This conversion looks suspicious, why do you need to build the host errors somewhere outside of the host?

Reply:
It's needed for `impl SnapshotSource for LedgerStorage`

Reply:
I see, makes sense.